// share/lib/server/browserchannel.js
var browserChannel, hat, syncQueue, util;

browserChannel = require('browserchannel').server;

util = require('util');

hat = require('hat');

syncQueue = require('./syncqueue');

module.exports = function(createClient, options) {
  options || (options = {});
  return browserChannel(options, function(session) {
    var buffer, bufferMsg, client, close, data, docState, handleClose, handleMessage, handleOp, handleOpenCreateSnapshot, lastReceivedDoc, lastSentDoc, open, send;
    data = {
      headers: session.headers,
      remoteAddress: session.address
    };
    client = null;
    lastSentDoc = null;
    lastReceivedDoc = null;
    docState = {};
    handleMessage = function(query) {
      var _name;
      if (query.doc === null) {
        query.doc = lastReceivedDoc = hat();
      } else if (query.doc !== void 0) {
        lastReceivedDoc = query.doc;
      } else {
        if (!lastReceivedDoc) {
          console.warn("msg.doc missing in query " + (JSON.stringify(query)) + " from " + client.id);
          return session.abort();
        }
        query.doc = lastReceivedDoc;
      }
      docState[_name = query.doc] || (docState[_name] = {
        queue: syncQueue(function(query, callback) {
          if (!docState) return callback();
          if (query.open === false) {
            return handleClose(query, callback);
          } else if (query.open || query.snapshot === null || query.create) {
            return handleOpenCreateSnapshot(query, callback);
          } else if (query.op != null) {
            return handleOp(query, callback);
          } else {
            console.warn("Invalid query " + (JSON.stringify(query)) + " from " + client.id);
            session.abort();
            return callback();
          }
        })
      });
      return docState[query.doc].queue(query);
    };
    send = function(response) {
      if (response.doc === lastSentDoc) {
        delete response.doc;
      } else {
        lastSentDoc = response.doc;
      }
      if (session.state !== 'closed') return session.send(response);
    };
    open = function(docName, version, callback) {
      var listener;
      if (!docState) return callback('Session closed');
      if (docState[docName].listener) return callback('Document already open');
      docState[docName].listener = listener = function(opData) {
        var opMsg;
        if (docState[docName].listener !== listener) {
          throw new Error('Consistency violation - doc listener invalid');
        }
        if (opData.meta.source === client.id) return;
        opMsg = {
          doc: docName,
          op: opData.op,
          v: opData.v,
          meta: opData.meta
        };
        return send(opMsg);
      };
      return client.listen(docName, version, listener, function(error, v) {
        if (error) delete docState[docName].listener;
        return callback(error, v);
      });
    };
    close = function(docName, callback) {
      var listener;
      if (!docState) return callback('Session closed');
      listener = docState[docName].listener;
      if (listener == null) return callback('Doc already closed');
      client.removeListener(docName);
      delete docState[docName].listener;
      return callback();
    };
    handleOpenCreateSnapshot = function(query, finished) {
      var callback, docData, docName, msg, step1Create, step2Snapshot, step3Open;
      docName = query.doc;
      msg = {
        doc: docName
      };
      callback = function(error) {
        if (error) {
          if (msg.open === true) close(docName);
          if (query.open === true) msg.open = false;
          if (query.snapshot !== void 0) msg.snapshot = null;
          delete msg.create;
          msg.error = error;
        }
        send(msg);
        return finished();
      };
      if (query.doc == null) return callback('No docName specified');
      if (query.create === true) {
        if (typeof query.type !== 'string') {
          return callback('create:true requires type specified');
        }
      }
      if (query.meta !== void 0) {
        if (!(typeof query.meta === 'object' && Array.isArray(query.meta) === false)) {
          return callback('meta must be an object');
        }
      }
      docData = void 0;
      /*
            client.getSnapshot query.doc, (error, data) ->
              maybeCreate = (callback) ->
                if query.create and error is 'Document does not exist'
                  client.create docName, query.type, query.meta or {}, callback
                else
                  callback error, data
      
              maybeCreate (error, data) ->
                if query.create
                  msg.create = !!error
                if error is 'Document already exists'
                  msg.create = false
                else if error and (!msg.create or error isnt 'Document already exists')
                  # This is the real final callback, to say an error has occurred.
                  return callback error
                else if query.create or query.snapshot is null
      
      
                if query.snapshot isnt null
      */
      step1Create = function() {
        if (query.create !== true) return step2Snapshot();
        if (docData) {
          msg.create = false;
          return step2Snapshot();
        } else {
          return client.create(docName, query.type, query.meta || {}, function(error) {
            if (error === 'Document already exists') {
              return client.getSnapshot(docName, function(error, data) {
                if (error) return callback(error);
                docData = data;
                msg.create = false;
                return step2Snapshot();
              });
            } else if (error) {
              return callback(error);
            } else {
              msg.create = !error;
              return step2Snapshot();
            }
          });
        }
      };
      step2Snapshot = function() {
        if (query.snapshot !== null || msg.create === true) {
          step3Open();
          return;
        }
        if (docData) {
          msg.v = docData.v;
          if (query.type !== docData.type.name) msg.type = docData.type.name;
          msg.snapshot = docData.snapshot;
        } else {
          return callback('Document does not exist');
        }
        return step3Open();
      };
      step3Open = function() {
        if (query.open !== true) return callback();
        if (query.type && docData && query.type !== docData.type.name) {
          return callback('Type mismatch');
        }
        return open(docName, query.v, function(error, version) {
          if (error) return callback(error);
          msg.open = true;
          msg.v = version;
          return callback();
        });
      };
      if (query.snapshot === null || (query.open === true && query.type)) {
        return client.getSnapshot(query.doc, function(error, data) {
          if (error && error !== 'Document does not exist') return callback(error);
          docData = data;
          return step1Create();
        });
      } else {
        return step1Create();
      }
    };
    handleClose = function(query, callback) {
      return close(query.doc, function(error) {
        if (error) {
          send({
            doc: query.doc,
            open: false,
            error: error
          });
        } else {
          send({
            doc: query.doc,
            open: false
          });
        }
        return callback();
      });
    };
    handleOp = function(query, callback) {
      var opData;
      opData = {
        v: query.v,
        op: query.op,
        meta: query.meta
      };
      return client.submitOp(query.doc, opData, function(error, appliedVersion) {
        var msg;
        msg = error ? {
          doc: query.doc,
          v: null,
          error: error
        } : {
          doc: query.doc,
          v: appliedVersion
        };
        send(msg);
        return callback();
      });
    };
    buffer = [];
    session.on('message', bufferMsg = function(msg) {
      return buffer.push(msg);
    });
    createClient(data, function(error, client_) {
      var msg, _i, _len;
      if (error) {
        session.send({
          auth: null,
          error: error
        });
        return session.stop();
      } else {
        client = client_;
        session.send({
          auth: client.id
        });
        session.removeListener('message', bufferMsg);
        for (_i = 0, _len = buffer.length; _i < _len; _i++) {
          msg = buffer[_i];
          handleMessage(msg);
        }
        buffer = null;
        return session.on('message', handleMessage);
      }
    });
    return session.on('close', function() {
      var docName, listener;
      if (!client) return;
      for (docName in docState) {
        listener = docState[docName].listener;
        if (listener) client.removeListener(docName);
      }
      return docState = null;
    });
  });
};
